### ISSUE TITLE
[Training] [M2] Integrate dataset with training pipeline

### ISSUE THREAD

**Issue by atfrank** [atfrank/MLFF-Distiller#17]
## Problem Statement

Ensure seamless integration between M2 dataset infrastructure and M1 training framework, validate training can begin immediately after M2.

## Context

- **Milestone**: M2 (Data Pipeline)
- **Priority**: High
- **Estimated Effort**: 2 days
- **Dependencies**: Issue #16 (Production Workflow)
- **Agent**: Distillation Training Engineer + Data Pipeline

## Acceptance Criteria

- [ ] Update training config to use production dataset
- [ ] Verify MolecularDataLoader works with HDF5 datasets
- [ ] Implement efficient data prefetching
- [ ] Test training loop with real data (no mock data)
- [ ] Benchmark data loading throughput (target: >90% GPU utilization)
- [ ] Validate normalization integration
- [ ] Create example training script using M2 dataset
- [ ] Documentation updates

## Deliverables

1. Updated training configuration files
2. `examples/train_with_production_data.py` - Example training script
3. Data loading benchmarks report
4. Integration tests
5. Documentation updates in `docs/training.md`

### 1. Configuration Updates

Update `configs/training_config.yaml`:
```yaml
data:
  dataset_path: "data/production_dataset_v1.0.h5"
  format: "hdf5"
  normalization_stats: "data/normalization_stats.yaml"
  train_split: "train"
  val_split: "val"
  test_split: "test"

dataloader:
  batch_size: 32
  num_workers: 4
  prefetch_factor: 2
  pin_memory: true
  persistent_workers: true
```

### 2. DataLoader Optimization

- Test num_workers (benchmark 1, 2, 4, 8 workers)
- Verify prefetching works correctly
- Test pin_memory for GPU transfer
- Monitor CPU/GPU utilization during training

### 3. Normalization Integration

Load normalization stats from M2 analysis:
```python
import yaml
with open('data/normalization_stats.yaml') as f:
    stats = yaml.safe_load(f)

# Apply in transforms
transforms = Compose([
    NormalizeEnergy(mean=stats['energy']['mean'], 
                   std=stats['energy']['std']),
    NormalizeForces(mean=stats['forces']['mean'],
                   std=stats['forces']['std']),
])
```

### 4. Training Validation

Run training for 100 iterations and verify:
- No data loading errors
- Batches contain correct data
- Normalization applied correctly
- Loss values reasonable
- GPU utilization >90%
- No memory leaks

## Benchmark Requirements

Create benchmark report with:
- Data loading throughput (samples/second)
- GPU utilization during training
- CPU utilization
- Memory usage (CPU and GPU)
- Comparison: num_workers 1 vs 4 vs 8

**Target**: >90% GPU utilization during training

## Example Training Script

Create `examples/train_with_production_data.py`:
```python
"""
Example training script using M2 production dataset.

Demonstrates:
- Loading HDF5 dataset
- Applying normalization
- Training loop with real data
- Validation loop
"""

from mlff_distiller.data import MolecularDataset, create_dataloaders
from mlff_distiller.training import DistillationTrainer
# ... rest of implementation
```

## Integration Tests

Create tests:
- [ ] Test HDF5 dataset loads correctly
- [ ] Test normalization applied correctly
- [ ] Test training loop runs without errors
- [ ] Test validation loop runs without errors
- [ ] Test dataloader with multiple workers
- [ ] Test pin_memory functionality

## Documentation Updates

Update `docs/training.md` with:
- How to use production dataset
- Normalization configuration
- DataLoader optimization tips
- Example training commands
- Troubleshooting common issues

## Technical Notes

- Use MolecularDataLoader from M1
- Verify augmentation pipeline compatible with HDF5
- Test with multiple workers (optimal num_workers)
- Profile data loading (identify bottlenecks using `mlff_distiller.cuda.md_profiler`)
- Ensure proper train/val/test split usage

## Performance Checklist

- [ ] Data loading not bottlenecking training (GPU >90% utilized)
- [ ] No data loading errors in 1000 iterations
- [ ] Normalization reducing loss faster than unnormalized
- [ ] Memory usage stable (no leaks)
- [ ] Prefetching working (no GPU stalls)

## Definition of Done

- [ ] Training successfully runs with production dataset
- [ ] GPU utilization >90% during training
- [ ] All integration tests passing
- [ ] Example script complete and tested
- [ ] Documentation updated
- [ ] Benchmark report complete
- [ ] M2 complete and ready for M3 (Model Architecture)
